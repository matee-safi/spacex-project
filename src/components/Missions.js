import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getMissions, joinMission } from '../redux/missions/missionsSlice';
import '../styles/Missions.css';

const Missions = () => {
  const dispatch = useDispatch();
  const {
    error, missionId, isPending, missionName, description, reserved,
  } = useSelector((state) => state.missions);

  useEffect(() => {
    dispatch(getMissions());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleJoin = (e) => {
    dispatch(joinMission(e.target.id));
  };

  return (
    <div className="missions">
      {isPending
      && (
      <div className="loading">
        <div className="sk-chase">
          <div className="sk-chase-dot" />
          <div className="sk-chase-dot" />
          <div className="sk-chase-dot" />
          <div className="sk-chase-dot" />
          <div className="sk-chase-dot" />
          <div className="sk-chase-dot" />
        </div>
      </div>
      )}
      {error && <p className="error">{error}</p>}
      {missionId.length > 0
      && (
      <table>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th> </th>
          </tr>
        </thead>
        <tbody>
          {missionId.map((mission, index) => (
            <tr key={mission}>
              <th className="mission-name"><p>{missionName[index]}</p></th>
              <td className="mission-description">{description[index]}</td>
              <td className="mission-status">{reserved[index] ? <p className="active-member">Active Member</p> : <p>NOT A MEMBER</p>}</td>
              <td className={reserved[index] ? 'leave-btn' : 'join-btn'}><button id={mission} type="button" onClick={(e) => handleJoin(e)}>{reserved[index] ? <>Leave Mission</> : <>Join Mission</>}</button></td>
            </tr>
          ))}
        </tbody>
      </table>
      )}
    </div>
  );
};

export default Missions;
