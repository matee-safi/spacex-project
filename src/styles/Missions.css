.missions {
  padding: 1em;
}

table {
  background-color: silver;
}

thead {
  text-align: left;
}

tr {
  background-color: white;
}

thead tr th {
  padding: 0.5em;
  background-color: white;
}

.mission-name {
  width: 10%;
  text-align: left;
  padding: 5px;
}

.mission-name p {
  width: 100%;
  height: 100%;
  margin: 0;
}

.mission-description {
  width: 65%;
  padding: 5px;
  padding-bottom: 2em;
}

.mission-status {
  width: 11%;
  text-align: center;
  padding: 5px;
}

.mission-status p {
  background-color: gray;
  color: white;
  font-weight: bold;
  border-radius: 5px;
  padding: 2px;
  margin: 0 auto;
}

.mission-btn {
  text-align: center;
}

.mission-btn button {
  background-color: transparent;
  font-size: 1em;
  font-weight: bolder;
  color: gray;
  border: 2px solid gray;
  padding: 5px 10px;
  cursor: pointer;
  transition: 0.2s;
}

.mission-btn button:hover {
  background-color: gray;
  color: white;
}

tr:nth-child(odd) {
  background-color: #f2f2f2;
}
