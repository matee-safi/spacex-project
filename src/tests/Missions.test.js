import renderer from 'react-test-renderer';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/Store';
import Missions from '../components/Missions';

it('renders Missions', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Missions />} />
          </Routes>
        </BrowserRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
