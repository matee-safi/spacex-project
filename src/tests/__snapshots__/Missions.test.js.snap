// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Missions 1`] = `
<div
  className="missions"
/>
`;
